### PR TITLE
Clean up particle-particle contact force class prior refactoring particle-wall contact force class in DEM

### PR DESCRIPTION
**Pull request by acdaigneault**
### Description

The particle-particle contact force class had some redundancies, so I managed to implement functions that are now called.
Some cleanup was been done here and there, but most of it is related to the call to the calculation of the contact force for a pair of particle to the proper model.
Before this refactoring, each (5) loop over the adjacent pair had the same series of `if constexpr (contact_model == ...)`, now each loop call `execute_contact_calculation()` according to the contact type as a template. 

This clean up was done before applying the same architecture pattern for the particle-wall contact force class.

### Testing

Tests are good.

### Miscellaneous (will be removed when merged)

I would like to have Gabo reviewing it before the merge since he has coded and may continue to code in the class and I want something that he agree on. However, I forbid him to do it during his mini-project hehe.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [x] If parameters are modified, the tests and the documentation of examples are up to date
- [x] Changelog (CHANGELOG.md) is up to date if the refactor affects the user experience or the codebase

Pull request related list:
- [x] No other PR is open related to this refactoring
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If any future works is planed, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/dem/dem_contact_manager.h [hunk 1]
```diff
@@ -219,15 +196,135 @@ class DEMContactManager
...
+  }
+
+  /**
+   * @brief Return the ghost particle-particle contact candidates.
+   */
+  inline typename dem_data_structures<dim>::particle_particle_candidates &
+  get_ghost_contact_pair_candidates()
+  {
+    return ghost_contact_pair_candidates;
+  }
```
Comment (acdaigneault) [inline include/dem/dem_contact_manager.h:325]: In order to keep the attributes private

File: include/dem/dem_contact_manager.h [hunk 2]
```diff
@@ -279,9 +376,9 @@ class DEMContactManager
     particle_floating_wall_in_contact;
   typename dem_data_structures<dim>::particle_wall_in_contact
     particle_wall_in_contact;
-  typename dem_data_structures<dim>::particle_line_in_contact
+  typename dem_data_structures<dim>::particle_line_candidates
     particle_lines_in_contact;
-  typename dem_data_structures<dim>::particle_point_in_contact
+  typename dem_data_structures<dim>::particle_point_candidates
     particle_points_in_contact;
```
Comment (acdaigneault) [inline include/dem/dem_contact_manager.h:382]: These were the same type

File: source/dem/particle_particle_contact_force.cc [hunk 3]
```diff
@@ -1,1284 +1,230 @@
...
+    typename dem_data_structures<dim>::adjacent_particle_pairs
+      &local_adjacent_particles,
+    typename dem_data_structures<dim>::adjacent_particle_pairs
+      &ghost_adjacent_particles,
+    typename dem_data_structures<dim>::adjacent_particle_pairs
+      &local_periodic_adjacent_particles,
+    typename dem_data_structures<dim>::adjacent_particle_pairs
+      &ghost_periodic_adjacent_particles,
+    typename dem_data_structures<dim>::adjacent_particle_pairs
+                              &ghost_local_periodic_adjacent_particles,
```
Comment (acdaigneault) [inline source/dem/particle_particle_contact_force.cc:33]: Do that (as it used to be before the implementation of the contact manager) so it is more clear what is neededin this function (also will be the same as for the pw contact force)

**Review by OGaboriault — COMMENTED** [review]
Still not done

File: include/dem/data_containers.h [hunk 4]
```diff
@@ -163,11 +163,14 @@ namespace DEM
                                    particle_wall_contact_info<dim>>>
       particle_wall_in_contact;
 
+    // <particle id, particle-particle info>
+    typedef ankerl::unordered_dense::map<types::particle_index,
+                                         particle_particle_contact_info<dim>>
+      particle_contact_info;
+
     // <particle id, <particle id, particle-particle info>>
```
[the remark this answers is not in the document]
Comment (acdaigneault) [reply include/dem/data_containers.h:171]: Thing is that could be j and i (ghost local pair)

File: include/dem/dem_contact_manager.h [hunk 5]
```diff
@@ -219,15 +196,135 @@ class DEMContactManager
...
+    return ghost_periodic_adjacent_particles;
+  }
+
+  /**
+   * @brief Return the ghost-local periodic particle contact container.
+   */
+  inline typename dem_data_structures<dim>::adjacent_particle_pairs &
+  get_ghost_local_periodic_adjacent_particles()
+  {
+    return ghost_local_periodic_adjacent_particles;
```
[the remark this answers is not in the document]
Comment (acdaigneault) [reply include/dem/dem_contact_manager.h:306]: It gets a bit long in my opinion, like the code gets heavy in that case, especially if we wanna say consistent and apply it on none periodic container. But feel free to do it :)

**Review by OGaboriault — COMMENTED** [review]
Not finish. Will rebase first

**Comment by blaisb** [comment]
@OGaboriault I think you should take over this PR. @acdaigneault has finished her master now so I'll you finalize it. I will review it over teh weekend.

**Comment by voferreira** [comment]
Is this ready for a review?

**Comment by acdaigneault** [comment]
Would be happy to awnser questions related to that PR, don't hesitate to tag me on some comments :)

**Review by voferreira — APPROVED** [review]
Nice refactoring! For me, it is good to go.

**Pull request closed without merging**